src/Common/tests/gtest_lru_resource_cache.cpp

```diff
@@ -0,0 +1,270 @@
+#include <iomanip>
+#include <iostream>
+#include <gtest/gtest.h>
+#include <Common/LRUResourceCache.h>
+
+TEST(LRUResourceCache, get)
+{
+    using MyCache = DB::LRUResourceCache<int, int>;
+    auto mcache = MyCache(10, 10);
+    int x = 10;
+    auto load_int = [&] { return std::make_shared<int>(x); };
+    auto holder1 = mcache.getOrSet(1, load_int);
+    x = 11;
+    auto holder2 = mcache.getOrSet(2, load_int);
+    ASSERT_TRUE(holder2 != nullptr);
+    ASSERT_TRUE(holder2->value() == 11);
+
+    auto holder3 = mcache.get(1);
+    ASSERT_TRUE(holder3 != nullptr);
+    ASSERT_TRUE(holder3->value() == 10);
+}
+
+TEST(LRUResourceCache, remove)
+{
+    using MyCache = DB::LRUResourceCache<int, int>;
+    auto mcache = MyCache(10, 10);
+    int x = 10;
+    auto load_int = [&] { return std::make_shared<int>(x); };
+    auto holder0 = mcache.getOrSet(1, load_int);
+    auto holder1 = mcache.getOrSet(1, load_int);
+
+    mcache.tryRemove(1);
+    holder0 = mcache.get(1);
+    ASSERT_TRUE(holder0 == nullptr);
+    auto n = mcache.size();
+    ASSERT_TRUE(n == 1);
+
+    holder0 = nullptr;
+    holder1 = nullptr;
+    n = mcache.size();
+    ASSERT_TRUE(n == 0);
+}
+
+struct MyWeight
+{
+    size_t operator()(const int & x) const { return static_cast<size_t>(x); }
+};
+
+TEST(LRUResourceCache, evictOnWweight)
+{
+    using MyCache = DB::LRUResourceCache<int, int, MyWeight>;
+    auto mcache = MyCache(5, 10);
+    int x = 2;
+    auto load_int = [&] { return std::make_shared<int>(x); };
+    auto holder1 = mcache.getOrSet(1, load_int);
+    holder1 = nullptr;
```

Comment:
Can also be like this
```suggestion
    holder1.reset();
```
(not necessary)

Reply:
done